README.md
# clj-scad-things

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar clj-scad-things-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-04-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-04-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clj-scad-things/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clj-scad-things/compare/0.1.0...0.1.1


test/clj_scad_things/core_test.clj
(ns clj-scad-things.core-test
  (:require [clojure.test :refer :all]
            [clj-scad-things.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clj_scad_things/misc.clj
(ns clj-scad-things.misc
  (:require [scad-clj.model :as m]
            [scad-clj.scad :as s]))

(def arduino-nano-handler
  (m/difference
    (m/union
      (m/minkowski
        (m/cube 44 18 3)
        (->> (m/cube 3 3 3)
             (m/rotate (/ (Math/PI) 4) [0 0 1])))

      (->> (m/cube 8 2 12)
           (m/translate [16.5 -10.2 6]))
      (->> (m/cube 8 2 12)
           (m/translate [-16.5 -10.2 6])))

    (m/minkowski
      (m/cube 40 14 3)
      (->> (m/cube 3 3 3)
           (m/rotate (/ (Math/PI) 4) [0 0 1])))


    (m/cube 44 18 2)
    (->> (m/cube 3 8 2)
         (m/translate [-23 0 2]))
    (->> (m/cylinder 1.8 2)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [-16.5 -10.2 8]))
    (->> (m/cylinder 1.8 2)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [16.5 -10.2 8]))

    (->> (m/cube 25 5 6)
         (m/translate [0 -10]))
    ))

(defn gen-arduino-nano-handler
  []
  (spit "resources/arduino-nano-handler.scad"
        (s/write-scad arduino-nano-handler)))

(def opto-handler
  (m/intersection
    (m/difference
      (m/cube 20 45 6)
      (->> (m/cube 10 14 6)
           (m/translate [5 16 0]))
      (->> (m/cube 10 14 6)
           (m/translate [5 -16 0]))
      (->> (m/cube 20 14.5 4)
           (m/translate [1.5 0 1]))
      (->> (m/cube 7 14 4)
           (m/translate [-5 16 1]))
      (->> (m/cube 7 14 4)
           (m/translate [-5 -16 1]))

      (->> (m/cylinder 1.8 6)
           (m/translate [-5 -12 0]))
      (->> (m/cylinder 1.8 6)
           (m/translate [-5 -18 0]))
      (->> (m/cube 3.4 6 6)
           (m/translate [-5 -15 0]))

      (->> (m/cylinder 1.8 6)
           (m/translate [-5 12 0]))
      (->> (m/cylinder 1.8 6)
           (m/translate [-5 18 0]))
      (->> (m/cube 3.4 6 6)
           (m/translate [-5 15 0]))


      (->> (m/cylinder 1.8 6)
           (m/translate [-3 0 0]))
      (->> (m/cylinder 1.8 6)
           (m/translate [5 0 0]))
      (->> (m/cube 8 3.2 6)
           (m/translate [1 0 0])))
    (->> (m/cube 20 50 50)
         (m/rotate (/ (Math/PI) 4) [1 0 0])
         (m/translate [0 0 -14]))))

(defn gen-opto-handler
  []
  (spit "resources/misc.opto-handler.scad"
        (s/write-scad opto-handler)))


(defn pi-handler
  []
  (letfn [(hex
            []
            (binding [m/*fn* 6]
              (m/cylinder 4 22)))]
    (m/difference
      (m/cube 20 70 22)
      (->> (m/cube 20 54 19)
           (m/translate [0 3 1.5]))
      (->> (m/cube 20 58 2)
           (m/translate [0 3 1]))

      (->> (m/cube 2.2 8 22)
           (m/translate [8.9 -31]))
      (->> (m/cube 2.2 8 22)
           (m/translate [-8.9 -31]))
      (->> (m/cube 3.2 8 22)
           (m/translate [3.5 -31]))
      (->> (m/cube 3.2 8 22)
           (m/translate [-3.5 -31]))

      (->> (m/cylinder 1.8 20)
           (m/rotate (/ (Math/PI) 2) [0 1 0])
           (m/translate [0 -31 1.5]))
      (->> (m/cylinder 1.8 20)
           (m/rotate (/ (Math/PI) 2) [0 1 0])
           (m/translate [0 -31 -6]))



      (->> (hex)
           (m/translate [-4 -15]))
      (->> (hex)
           (m/translate [-4 -5]))
      (->> (hex)
           (m/translate [-4 5]))
      (->> (hex)
           (m/translate [-4 15]))
      (->> (hex)
           (m/translate [-4 25]))

      (->> (hex)
           (m/translate [4 -10]))
      (->> (hex)
           (m/translate [4 0]))
      (->> (hex)
           (m/translate [4 10]))
      (->> (hex)
           (m/translate [4 20]))

      (->> (m/cube 12 13 9)
           (m/translate [0 -28.5 9.5]))
      (->> (m/cube 20 3 10)
           (m/rotate (/ (Math/PI) 12) [1 0 0])
           (m/translate [0 -24 7]))

      (->> (m/cube 16 5 7)
           (m/translate [0 32.5 5.5]))
      (->> (m/cube 20 3 10)
           (m/rotate (/ (Math/PI) -12) [1 0 0])
           (m/translate [0 30 7])))))

(defn gen-pi-handler
  []
  (spit "resources/misc.pi-handler.scad"
        (s/write-scad (pi-handler))))

(defn cover-1
  []
  (m/difference
    (m/cube 56 30 17)
    (->> (m/cube 42 31 2)
         (m/rotate (/ (Math/PI) 96) [1 0 0])
         (m/translate [0 0 9]))
    (->> (m/cube 10 30 40)
         (m/rotate (/ (Math/PI) 6) [0 1 0])
         (m/translate [-29 0 0]))
    (->> (m/cube 10 30 40)
         (m/rotate (/ (Math/PI) -6) [0 1 0])
         (m/translate [29 0 0]))
    (->> (m/cube 19.5 30 13)
         (m/translate [0 0 -2]))
    (->> (m/cube 22.8 30 6)
         (m/translate [0 0 -6]))

    (->> (m/cylinder 2.6 56)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [0 -7 -1]))
    (->> (m/cylinder 6 16)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [22 -7 -1]))
    (->> (m/cylinder 6 16)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [-22 -7 -1]))

    (->> (m/cube 19.5 10 17)
         (m/translate [0 10 0]))
    (->> (m/cube 22.8 6 8)
         (m/translate [0 12 0]))))

(defn gen-cover-1
  []
  (spit "resources/misc.cover-1.scad"
        (s/write-scad (cover-1))))

src/clj_scad_things/pulley.clj
(ns clj-scad-things.pulley
  (:require [scad-clj.model :as m]
            [scad-clj.scad :as s]))


(defn rpc-90
  []
  (as-> (m/polygon [[10 0]
                    [10 7.25]
                    [20 7.25]
                    [20 12.5]
                    [43 12.5]
                    [45 10.5]
                    [45 -10.5]
                    [43 -12.5]
                    [20 -12.5]
                    [20 -7.25]
                    [10 -7.25]]) $
        (binding [m/*fn* 256]
          (m/extrude-rotate $))))

(defn gen-rpc-90
  []
  (spit "resources/pulley/rpc-90.scad"
        (s/write-scad (rpc-90))))

(defn r-33
  []
  (as-> (m/polygon [[5 4]
                    [12 4]
                    [12 5.5]
                    [15 5.5]
                    [16.5 4]
                    [16.5 -4]
                    [15 -5.5]
                    [12 -5.5]
                    [12 -4]
                    [5 -4]]) $
        (binding [m/*fn* 256]
          (m/extrude-rotate $))))

(defn gen-r-33
  []
  (spit "resources/pulley/r-33.scad"
        (s/write-scad (r-33))))

(defn arm
  []
  (m/difference
    (m/cube 145 120 40)
    (->> (m/cube 142 114 40)
         (m/translate [-1.5 0 0]))
    (->> (m/cube 3 120 20)
         (m/translate [71 0 -10]))
    (->> (m/cube 4 120 10)
         (m/translate [67.5 0 -5]))
    (->> (m/cylinder 11.5 3)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [-52.5 -58.5 0]))
    (->> (m/cylinder 8.2 3)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [-52.5 58.5 0]))

    (->> (m/cylinder 1.8 120)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [-36.9 0 -15.6]))
    (->> (m/cylinder 1.8 120)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [-36.9 0 15.6]))
    (->> (m/cylinder 1.8 120)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [-68.1 0 -15.6]))
    (->> (m/cylinder 1.8 120)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [-68.1 0 15.6]))))

(defn gen-arm
  []
  (spit "resources/pulley/arm.scad"
        (s/write-scad (arm))))

(defn support
  []
  (m/difference
    (m/cube 45 126 40)
    (->> (m/cube 42 120 40)
         (m/translate [1.5 0 0]))
    (->> (m/cylinder 11.5 3)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [2.5 -61.5 0]))
    (->> (m/cylinder 8.2 3)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [2.5 61.5 0]))

    (->> (m/cylinder 1.8 126)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [-13.1 0 -15.6]))
    (->> (m/cylinder 1.8 126)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [-13.1 0 15.6]))
    (->> (m/cylinder 1.8 126)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [18.1 0 -15.6]))
    (->> (m/cylinder 1.8 126)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [18.1 0 15.6]))

    (->> (m/cylinder 2.3 3)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [-21 -17.5 10]))
    (->> (m/cylinder 2.3 3)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [-21 17.5 10]))
    (->> (m/cylinder 2.3 3)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [-21 -17.5 -10]))
    (->> (m/cylinder 2.3 3)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [-21 17.5 -10]))))

(defn gen-support
  []
  (spit "resources/pulley/support.scad"
        (s/write-scad (support))))

(defn chassis-partial
  []
  (m/difference
    (m/cube 180 192 38)
    (->> (m/cube 177 192 35)
         (m/translate [-1.5 0 1.5]))

    (->> (m/cube 50 33 38)
         (m/translate [65 83.5 0]))
    (->> (m/cube 50 33 38)
         (m/translate [65 -83.5 0]))
    (->> (m/cube 106 33 38)
         (m/translate [-37 83.5 0]))
    (->> (m/cube 106 33 38)
         (m/translate [-37 -83.5 0]))

    (->> (m/cube 4 15 3)
         (m/translate [17.335 83.5 -17.5]))
    (->> (m/cube 4 15 3)
         (m/translate [38.665 83.5 -17.5]))
    (->> (m/cube 4 15 3)
         (m/translate [17.335 -83.5 -17.5]))
    (->> (m/cube 4 15 3)
         (m/translate [38.665 -83.5 -17.5]))

    (->>
      (m/difference
        (m/cylinder 15 20)
        (m/cylinder 10 20))
      (m/rotate (/ (Math/PI) 2) [1 0 0])
      (m/translate [28 96 -17.5]))
    (->>
      (m/difference
        (m/cylinder 15 20)
        (m/cylinder 10 20))
      (m/rotate (/ (Math/PI) 2) [1 0 0])
      (m/translate [28 -96 -17.5]))

    (->> (m/cube 3 5 12)
         (m/translate [88.5 57.5 -6]))
    (->> (m/cube 3 5 12)
         (m/translate [88.5 -57.5 -6]))
    (->> (m/cube 3 5 12)
         (m/translate [88.5 57.5 14]))
    (->> (m/cube 3 5 12)
         (m/translate [88.5 -57.5 14]))


    (->> (m/cylinder 4.5 3)
         (m/translate [58 50 -17.5]))
    (->> (m/cylinder 4.5 3)
         (m/translate [58 -50 -17.5]))

    (->> (binding [m/*fn* 6]
           (m/cylinder 50 3))
         (m/translate [15 0 -17.5]))
    (->> (binding [m/*fn* 6]
           (m/cylinder 25 3))
         (m/translate [-55 -30 -17.5]))
    (->> (binding [m/*fn* 6]
           (m/cylinder 25 3))
         (m/translate [-55 30 -17.5]))))

(defn gen-chassis-partial
  []
  (spit "resources/pulley/chassis-partial.scad"
        (s/write-scad (chassis-partial))))

(defn pulley-unit
  []
  (m/difference
    (m/cube 64 192 48)
    (->> (m/cube 56 192 44)
         (m/translate [0 0 2]))

    (->> (m/cube 50 33 48)
         (m/translate [0 83.5 0]))
    (->> (m/cube 50 33 48)
         (m/translate [0 -83.5 0]))


    (->> (m/cube 64 192 48)
         (m/translate [-4 0 16]))


    (->> (m/cube 4 15 6)
         (m/translate [-30 83.5 -9.5]))
    (->> (m/cube 4 15 6)
         (m/translate [-30 -83.5 -9.5]))

    (->> (m/cylinder 4.5 4)
         (m/translate [0 50 -22]))
    (->> (m/cylinder 4.5 4)
         (m/translate [0 -50 -22]))

    (->> (m/difference
           (m/cylinder 100 16.2)
           (m/cylinder 4.5 16.2))
         (m/translate [30 -83.5 16]))
    (->> (m/difference
           (m/cylinder 100 16.2)
           (m/cylinder 4.5 16.2))
         (m/translate [30 83.5 16]))

    (->> (m/cube 10 10 48)
         (m/translate [30 96 0]))
    (->> (m/cube 10 10 48)
         (m/translate [30 -96 0]))
    (->> (m/cube 10 150 48)
         (m/translate [30 0 24]))

    (->> (m/cylinder 1.5 4)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [30 -83.5 20]))
    (->> (m/cylinder 1.5 4)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [30 83.5 20]))))

(defn gen-pulley-unit
  []
  (spit "resources/pulley/pulley-unit.scad"
        (s/write-scad (pulley-unit))))



src/clj_scad_things/dr_24312.clj
(ns clj-scad-things.dr-24312
  (:require [scad-clj.model :as m]
            [scad-clj.scad :as s]))

(def dot-projector
  (m/difference
    (m/union
      (m/minkowski
        (->> (m/cube 7 7 5)
             (m/rotate (/ (Math/PI) 4) [0 0 1]))
        (m/cube 30 30 10)))
    (m/cylinder 8 19)
    (->> (m/cube 40 32 1)
         (m/translate [0 0 7]))
    (->> (m/cube 32 20 2)
         (m/translate [-4 -6 5.5]))
    (->> (m/cube 5 5 6)
         (m/translate [17.5 0 3.5]))
    (->> (m/cylinder 16 14)
         (m/translate [0 0 0.5]))

    (->> (m/cylinder 1.9 15)
         (m/translate [16.5 11]))
    (->> (m/cylinder 3 14)
         (m/translate [16.5 11 -5]))
    (->> (m/cube 6 6 14)
         (m/translate [19.7 11 -5]))

    (->> (m/cylinder 1.9 15)
         (m/translate [16.5 -11]))
    (->> (m/cylinder 3 14)
         (m/translate [16.5 -11 -5]))
    (->> (m/cube 6 6 14)
         (m/translate [19.7 -11 -5]))

    (->> (m/cylinder 1.9 15)
         (m/translate [-17.5 11]))
    (->> (m/cylinder 3 14)
         (m/translate [-17.5 11 -5]))
    (->> (m/cube 6 6 14)
         (m/translate [-20.7 11 -5]))

    (->> (m/cube 15 6 4)
         (m/translate [-20 -6 -4])
         (m/rotate (/ (Math/PI) 6) [0 0 1]))
    (->> (m/cube 6 14 2)
         (m/translate [-20 -6 -4])
         (m/rotate (/ (Math/PI) 6) [0 0 1]))
    (->> (m/cube 6 2 10)
         (m/translate [-20 -6 -4])
         (m/rotate (/ (Math/PI) 6) [0 0 1]))
    (->> (m/cylinder 3.2 5.5)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [-20 -6 -4])
         (m/rotate (/ (Math/PI) 6) [0 0 1]))
    (->> (m/cube 6 15 15)
         (m/translate [-25.5 -6])
         (m/rotate (/ (Math/PI) 6) [0 0 1]))

    (->> (m/cube 10 10 4)
         (m/translate [-6 -16 -4]))))

(defn gen-dot-projector
  []
  (spit "resources/dr_24312/dot-projector.scad"
        (s/write-scad dot-projector)))



(def line-pointer-handle
  (m/difference
    (m/union
      (m/cylinder 23 2)
      (->> (m/cube 18 2 10)
           (m/translate [24 0 4])
           (m/rotate (/ (Math/PI) -6) [0 0 1]))
      (->> (m/intersection
             (->> (m/cube 2 12 12)
                  (m/rotate (/ (Math/PI) 4) [1 0 0]))
             (->> (m/cube 2 30 30)
                  (m/translate [0 0 15])))
           (m/translate [18 0 0])
           (m/rotate (/ (Math/PI) -6) [0 0 1])

           ))
    (m/cylinder 15 2)

    (->> (m/cylinder 1.8 2)
         (m/translate [-16.5 11]))

    (->> (m/cylinder 1.8 2)
         (m/translate [-16.5 -11]))

    (->> (m/cylinder 1.8 2)
         (m/translate [17.5 11]))

    (->> (m/cylinder 2.2 3)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [28 0 4])
         (m/rotate (/ (Math/PI) -6) [0 0 1]))))

(defn gen-line-pointer-handle
  []
  (spit "resources/dr_24312/line-pointer-handle.scad"
        (s/write-scad line-pointer-handle)))


(def dot-projector-handle
  (m/difference
    (m/union
      (m/minkowski
        (->> (m/cube 7 7 5)
             (m/rotate (/ (Math/PI) 4) [0 0 1]))
        (m/cube 30 30 5)))
    (->> (m/cube 18 40 8)
         (m/translate [11 0 1]))
    (->> (m/cube 18 40 8)
         (m/translate [-11 0 1]))
    (->> (m/cylinder 6 2)
         (m/translate [0 0 -4]))
    (m/cube 4 10 10)

    (->> (m/cylinder 1.8 10)
         (m/translate [16.5 11]))
    (->> (m/cylinder 1.8 10)
         (m/translate [16.5 -11]))
    (->> (m/cylinder 1.8 10)
         (m/translate [-17.5 11]))

    (->> (m/cylinder 1.8 4)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [0 15 1]))
    (->> (m/cylinder 1.8 4)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [0 -15 1]))
    (->> (m/cylinder 1.8 4)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [0 8 1]))
    (->> (m/cylinder 1.8 4)
         (m/rotate (/ (Math/PI) 2) [0 1 0])
         (m/translate [0 -8 1]))))

(defn gen-dot-projector-handle
  []
  (spit "resources/dr_24312/dot-projector-handle.scad"
        (s/write-scad dot-projector-handle)))


(defn pointer-head
  []
  (binding [m/*fn* 30]
    (m/difference
      (m/sphere 6)
      (->> (m/cube 12 12 2)
           (m/translate [0 0 5]))
      (->> (m/cube 12 12 2)
           (m/translate [0 0 -5]))
      (m/cylinder 3.1 12)
      (m/cube 3 7 12))))

(defn gen-pointer-head
  []
  (spit "resources/dr-24312/pointer-head.scad"
        (s/write-scad (pointer-head))))

(defn pointer-unit
  []
  (m/difference
    (m/union
      (m/cube 48 44 2)
      (->> (m/cube 48 8 4)
           (m/translate [0 -18 3]))
      (->> (m/cylinder 4 4)
           (m/translate [0 2 3]))
      (->> (m/cube 6 8 4)
           (m/translate [0 18 3]))
      (->> (m/cube 8 28 6)
           (m/translate [-24 -8 2]))
      (->> (m/cube 8 28 6)
           (m/translate [24 -8 2])))
    (->> (m/cylinder 3.2 4)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [0 -20 5.5]))
    (->> (m/cube 8 8 3)
         (m/translate [-10 -18 3.5]))
    (->> (m/cube 8 8 3)
         (m/translate [10 -18 3.5]))

    (->> (m/sphere 6)
         (m/translate [-18 16 5.5]))
    (->> (m/sphere 6)
         (m/translate [-12 4 5.5]))
    (->> (m/sphere 6)
         (m/translate [-6 -8 5.5]))
    (->> (m/sphere 6)
         (m/translate [18 16 5.5]))
    (->> (m/sphere 6)
         (m/translate [12 4 5.5]))
    (->> (m/sphere 6)
         (m/translate [6 -8 5.5]))

    (->> (m/cube 16 4 10)
         (m/translate [0 -16]))
    (->> (m/cube 20 2 10)
         (m/translate [0 -15]))

    (->> (m/cube 32 1 10)
         (m/translate [0 -2]))
    (->> (m/cube 1 12 10)
         (m/translate [0 -8]))
    (->> (m/cube 1 10 10)
         (m/translate [15.5 -7]))
    (->> (m/cube 1 10 10)
         (m/translate [-15.5 -7]))

    (->> (m/cube 1 18 10)
         (m/translate [-20 1]))
    (->> (m/cube 14 1 10)
         (m/translate [-13 9.5]))
    (->> (m/cube 1 8 10)
         (m/translate [-5.5 6]))
    (->> (m/cube 2 1 10)
         (m/translate [-5 2]))

    (->> (m/cube 1 18 10)
         (m/translate [20 1]))
    (->> (m/cube 14 1 10)
         (m/translate [13 9.5]))
    (->> (m/cube 1 8 10)
         (m/translate [5.5 6]))
    (->> (m/cube 2 1 10)
         (m/translate [5 2]))


    (->> (m/cube 1 9 10)
         (m/translate [-3.5 17.5]))
    (->> (m/cube 1 2 10)
         (m/translate [-2.5 13]))
    (->> (m/cube 1 9 10)
         (m/translate [3.5 17.5]))
    (->> (m/cube 1 2 10)
         (m/translate [2.5 13]))
    (->> (m/cylinder 3.1 8)
         (m/rotate (/ (Math/PI) 2) [1 0 0])
         (m/translate [0 18 5.5]))


    (->> (m/cylinder 1.8 10)
         (m/translate [-24 2]))
  (->> (m/cylinder 1.8 10)
       (m/translate [24 2]))
    (->> (m/cylinder 1.8 10)
         (m/translate [0 2]))
    (->> (m/cylinder 1.8 10)
         (m/translate [-20 -18]))
    (->> (m/cylinder 1.8 10)
         (m/translate [20 -18]))))

(defn gen-pointer-unit
  []
  (spit "resources/dr-24312/pointer-unit.scad"
        (s/write-scad (pointer-unit))))




src/clj_scad_things/honeycomb.clj
(ns clj-scad-things.honeycomb
  (:require [scad-clj.model :as m]
            [scad-clj.scad :as s]))

(defn stamp
  [r h]
  (let [sin60 (Math/sin (/ (Math/PI) 3))
        cos60 (Math/cos (/ (Math/PI) 3))]
    (m/intersection
      (->> (m/polygon [[(* 0.1 cos60) (* 0.1 sin60)]
                       [-0.1 0]
                       [(* 0.1 cos60) (* -0.1 sin60)]])
           (m/extrude-linear {:height h
                              :scale  (/ (* 10 h) (Math/tan (/ (Math/PI) 12)))
                              :center false}))
      (->> (m/polygon [[r 0]
                       [(* r cos60) (* r sin60)]
                       [(- (* r cos60)) (* r sin60)]
                       [(- r) 0]
                       [(- (* r cos60)) (- (* r sin60))]
                       [(* r cos60) (* (- r) sin60)]])
           (m/extrude-linear {:height h
                              :center false})))))

(defn cell
  [r h d]
  (m/difference
    (stamp (+ r (/ d 2)) h)
    (->> (stamp (- r (/ d 2)) h)
         (m/translate [0 0 d]))))

(defn cells
  [xn yn r h d]
  (m/union
    (->> (for [i (range xn)
               j (range yn)
               :when (= (even? i) (even? j))]
           (->> (cell r h d)
                (m/translate [(* 1.5 i r) (/ (* j (Math/sqrt 3) r) 2) 0]))))))



(defn gen-cells
  []
  (spit "resources/honeycomb/test.scad"
        (s/write-scad
          (->> (cells 25 15 10 10 1)
               (m/rotate (/ (Math/PI) 2) [1 0 0])))))

src/clj_scad_things/rover.clj
(ns clj-scad-things.rover
  (:require [scad-clj.model :as m]
            [scad-clj.scad :as s]))


(defn track-drive-wheel
  []
  (binding [m/*fn* 256]
    (m/difference
      (m/union
        (m/cylinder 29.5 10)
        (->> (m/cylinder 24 12)
             (m/translate [0 0 11])))
      (m/cylinder 18 10)
      (->> (m/cylinder 19.8 12)
           (m/translate [0 0 11]))
      (for [i (range 12)]
        (->> (binding [m/*fn* 64]
               (m/cylinder 3.5 11))
             (m/translate [29.5 0 0])
             (m/rotate (* (/ (Math/PI) 6) i) [0 0 1]))))))

(defn gen-track-drive-wheel
  []
  (spit "resources/rover/track-drive-wheel.scad"
        (s/write-scad (track-drive-wheel))))


(defn drive-wheel
  []
  (m/difference
    (m/union
      (as-> (m/polygon [[2.5 0]
                        [2.5 22]
                        [6 22]
                        [8 8]
                        [41 8]
                        [41 50]
                        [45 50]
                        [45 49]
                        [44 48]
                        [44 31]
                        [45 30]
                        [45 0]]) $
            (binding [m/*fn* 256]
              (m/extrude-rotate $)))
      (for [i (range 8)]
        (->> (m/polygon [[0 0]
                         [0 50]
                         [5 0]])
             (m/extrude-linear {:height    3
                                :convexity 10})
             (m/rotate (/ (Math/PI) 2) [1 0 0])
             (m/translate [-41 0 0])
             (m/rotate (* i (/ (Math/PI) 4)) [0 0 1])))
      (for [i (range 8)]
        (->> (m/polygon [[0 0]
                         [0 25]
                         [25 0]])
             (m/extrude-linear {:height    3
                                :convexity 10})
             (m/rotate (/ (Math/PI) 2) [1 0 0])
             (m/translate [-41 0 0])
             (m/rotate (+ (* i (/ (Math/PI) 4)) (/ (Math/PI) 8)) [0 0 1]))))

    (m/cube 1 30 44)
    (m/cube 30 1 44)

    (for [i (range 8)]
      (->> (m/cylinder 8 16)
           (m/translate [26 0 0])
           (m/rotate (* i (/ (Math/PI) 4)) [0 0 1])))
    (apply concat
           (for [x [-8 8]
                 y [-8 8]]

             [(->> (m/cylinder 1.8 10)
                   (m/translate [x y 5]))
              (->> (m/cylinder 4 3)
                   (m/translate [x y 1.5]))]))
    (->> (m/difference
           (m/cylinder 38 2)
           (m/cylinder 36 2))
         (m/translate [0 0 1]))))

(defn gen-drive-wheel
  []
  (spit "resources/rover/drive-wheel.scad"
        (s/write-scad (drive-wheel))))

(defn rear-wheel
  []
  (m/difference
    (m/union
      (as-> (m/polygon [[8.25 0]
                        [8.25 15]
                        [7 16]
                        [7 44]
                        [8.25 45]
                        [8.25 50]
                        [12 50]
                        [12 4]
                        [42 4]
                        [42 50]
                        [45 50]
                        [45 49]
                        [44 48]
                        [44 31]
                        [45 30]
                        [45 0]]) $
            (binding [m/*fn* 256]
              (m/extrude-rotate $)))
      (for [i (range 8)]
        (->> (m/cube 32 48 2)
             (m/rotate (/ (Math/PI) 2) [1 0 0])
             (m/translate [-27 0 26])
             (m/rotate (+ (* i (/ (Math/PI) 4)) (/ (Math/PI) 8)) [0 0 1]))))

    (for [i (range 8)]
      (->> (m/cylinder 8 16)
           (m/translate [26 0 0])
           (m/rotate (* i (/ (Math/PI) 4)) [0 0 1])))
    (->> (m/difference
           (m/cylinder 38 2)
           (m/cylinder 36 2))
         (m/translate [0 0 1]))))

(defn gen-rear-wheel
  []
  (spit "resources/rover/rear-wheel.scad"
        (s/write-scad (rear-wheel))))

(defn steering
  []
  (letfn [(cyl [r h n]
            (binding [m/*fn* n]
              (m/cylinder r h)))]
    (m/difference
      (m/union
        (m/cylinder 30 16)
        (->> (m/cylinder 32 2)
             (m/translate [0 0 7]))
        (->> (m/cylinder 32 2)
             (m/translate [0 0 -2]))
        (for [i (range -80 80 2)
              :let [a (+ (/ (inc i) 30) (/ (Math/PI) 2))]]
          (->> (cyl 0.5 7 32)
               (m/translate [30 0 2.5])
               (m/rotate a [0 0 1]))))
      (->> (m/cube 43 40.5 12)
           (m/translate [0 0 2.5]))
      (for [i (range -80 80 2)
            :let [a (+ (/ i 30) (/ (Math/PI) 2))]]
        (->> (cyl 0.55 7 32)
             (m/translate [30 0 2.5])
             (m/rotate a [0 0 1])))

      (->> (m/cylinder 1.8 10)
           (m/rotate (/ (Math/PI) 2) [1 0 0])
           (m/translate [15.5 -25 2.5]))
      (->> (m/cylinder 1.8 10)
           (m/rotate (/ (Math/PI) 2) [1 0 0])
           (m/translate [-15.5 -25 2.5]))


      (->> (m/cylinder 1.8 16)
           (m/translate [0 25 0]))
      (->> (m/cylinder 3 10)
           (m/translate [0 25 0]))
      (->> (m/cylinder 1.8 16)
           (m/translate [25 0 0]))
      (->> (m/cylinder 3 10)
           (m/translate [25 0 0]))
      (->> (m/cylinder 1.8 16)
           (m/translate [-25 0 0]))
      (->> (m/cylinder 3 10)
           (m/translate [-25 0 0]))

      #_(m/cylinder 7 16)
      (->> (cyl 8.25 5 64)
           (m/translate [0 0 -5.5]))
      (->> (m/difference
             (m/cylinder 32 3)
             (m/cylinder 10 3))
           (m/translate [0 0 -6.5]))

      (->> (m/cube 60 10 16)
           (m/translate [0 -30 0]))

      #_(->> (m/cube 60 10 10)
             (m/translate [0 30 0]))
      #_(->> (m/cube 40 4 16)
             (m/translate [0 30 0]))

      (->> (m/cylinder 15 10)
           (m/rotate (/ (Math/PI) 2) [1 0 0])
           (m/translate [0 -25 19])))))

(defn gen-steering
  []
  (spit "resources/rover/steering.scad"
        (s/write-scad (steering))))